File: client-sdk-python/amttp/services/transaction.py
"""Transaction Service — wraps ``/tx`` endpoints."""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel

from amttp.services.base import BaseService

# ── types ─────────────────────────────────────────────────────────────────────

TransactionStatus = Literal["pending", "processing", "completed", "failed", "cancelled"]


class TransactionRequest(BaseModel):
    from_address: str
    to_address: str
    amount: str
    token_address: Optional[str] = None
    chain_id: int = 1
    memo: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None


class TransactionRecord(BaseModel):
    id: str
    hash: Optional[str] = None
    from_address: str = ""
    to_address: str = ""
    amount: str = ""
    token_address: Optional[str] = None
    chain_id: int = 1
    status: str = "pending"
    risk_score: Optional[float] = None
    risk_level: Optional[str] = None
    policy_result: Optional[Dict[str, Any]] = None
    created_at: str = ""
    updated_at: str = ""
    completed_at: Optional[str] = None
    memo: Optional[str] = None


class TransactionValidation(BaseModel):
    valid: bool = False
    risk_score: float = 0
    risk_level: str = "low"
    policy_result: Optional[Dict[str, Any]] = None
    label_warnings: List[Dict[str, Any]] = []
    estimated_gas: Optional[str] = None


# ── service ───────────────────────────────────────────────────────────────────


class TransactionService(BaseService):
    """Transaction management operations."""

    async def validate(self, request: TransactionRequest) -> TransactionValidation:
        """Validate a transaction before submission."""
        resp = await self._http.post(
            "/tx/validate",
            json=request.model_dump(exclude_none=True, by_alias=True),
        )
        resp.raise_for_status()
        data = TransactionValidation.model_validate(resp.json())
        self._events.emit("transaction:validated", data.model_dump())
        return data

    async def submit(self, request: TransactionRequest) -> TransactionRecord:
        """Submit a new transaction."""
        resp = await self._http.post(
            "/tx/submit",
            json=request.model_dump(exclude_none=True, by_alias=True),
        )
        resp.raise_for_status()
        data = TransactionRecord.model_validate(resp.json())
        self._events.emit("transaction:submitted", {"id": data.id, "hash": data.hash})
        return data

    async def get_status(self, tx_id: str) -> TransactionRecord:
        """Get transaction status."""
        resp = await self._http.get(f"/tx/{tx_id}")
        resp.raise_for_status()
        return TransactionRecord.model_validate(resp.json())

    async def get_history(
        self,
        address: str,
        *,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
        status: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Get transaction history for an address."""
        params: Dict[str, str] = {}
        if limit is not None:
            params["limit"] = str(limit)
        if offset is not None:
            params["offset"] = str(offset)
        if status:
            params["status"] = status
        resp = await self._http.get(f"/tx/history/{address}", params=params)
        resp.raise_for_status()
        return resp.json()

    async def cancel(self, tx_id: str) -> TransactionRecord:
        """Cancel a pending transaction."""
        resp = await self._http.post(f"/tx/{tx_id}/cancel")
        resp.raise_for_status()
        data = TransactionRecord.model_validate(resp.json())
        self._events.emit("transaction:cancelled", {"id": data.id})
        return data

    async def retry(self, tx_id: str) -> TransactionRecord:
        """Retry a failed transaction."""
        resp = await self._http.post(f"/tx/{tx_id}/retry")
        resp.raise_for_status()
        data = TransactionRecord.model_validate(resp.json())
        self._events.emit("transaction:retried", {"id": data.id})
        return data

File: client-sdk/src/services/transaction.ts
/**
 * Transaction Service
 * Wraps /tx endpoints for transaction management
 */

import { AxiosInstance } from 'axios';
import { EventEmitter } from '../events';
import { BaseService } from './base';

// Local type definitions
export type TransactionStatus = 'pending' | 'processing' | 'completed' | 'failed' | 'cancelled';
export type RiskLevel = 'low' | 'medium' | 'high' | 'critical';

export interface TransactionRequest {
  from: string;
  to: string;
  amount: string;
  tokenAddress?: string;
  chainId: number;
  memo?: string;
  metadata?: Record<string, any>;
}

export interface TransactionRecord {
  id: string;
  hash?: string;
  from: string;
  to: string;
  amount: string;
  tokenAddress?: string;
  chainId: number;
  status: TransactionStatus;
  riskScore?: number;
  riskLevel?: RiskLevel;
  policyResult?: {
    allowed: boolean;
    reason?: string;
    appliedPolicies: string[];
  };
  createdAt: string;
  updatedAt: string;
  completedAt?: string;
  memo?: string;
}

export interface TransactionValidation {
  valid: boolean;
  riskScore: number;
  riskLevel: RiskLevel;
  policyResult: {
    allowed: boolean;
    reason?: string;
    appliedPolicies: string[];
    requiredApprovals?: number;
  };
  labelWarnings: {
    address: string;
    labels: string[];
    severity: 'low' | 'medium' | 'high' | 'critical';
  }[];
  estimatedGas?: string;
  estimatedFee?: string;
}

export interface TransactionHistoryOptions {
  address?: string;
  status?: TransactionStatus;
  startDate?: Date;
  endDate?: Date;
  limit?: number;
  offset?: number;
  sortBy?: 'createdAt' | 'amount' | 'riskScore';
  sortOrder?: 'asc' | 'desc';
}

export class TransactionService extends BaseService {
  constructor(http: AxiosInstance, events: EventEmitter) {
    super(http, events);
  }

  /**
   * Validate a transaction before submission
   */
  async validate(request: TransactionRequest): Promise<TransactionValidation> {
    const response = await this.http.post<TransactionValidation>('/tx/validate', request);
    
    this.events.emit('transaction:validated', {
      from: request.from,
      to: request.to,
      valid: response.data.valid,
      riskLevel: response.data.riskLevel
    });

    return response.data;
  }

  /**
   * Submit a transaction for processing
   */
  async submit(request: TransactionRequest): Promise<TransactionRecord> {
    const response = await this.http.post<TransactionRecord>('/tx/submit', request);
    
    this.events.emit('transaction:submitted', {
      id: response.data.id,
      from: request.from,
      to: request.to,
      status: response.data.status
    });

    return response.data;
  }

  /**
   * Get transaction by ID
   */
  async get(id: string): Promise<TransactionRecord> {
    const response = await this.http.get<TransactionRecord>(`/tx/${id}`);
    return response.data;
  }

  /**
   * Get transaction by hash
   */
  async getByHash(hash: string): Promise<TransactionRecord> {
    const response = await this.http.get<TransactionRecord>(`/tx/hash/${hash}`);
    return response.data;
  }

  /**
   * Get transaction history
   */
  async getHistory(options?: TransactionHistoryOptions): Promise<{
    transactions: TransactionRecord[];
    total: number;
    hasMore: boolean;
  }> {
    const params = new URLSearchParams();
    if (options?.address) params.append('address', options.address);
    if (options?.status) params.append('status', options.status);
    if (options?.startDate) params.append('startDate', options.startDate.toISOString());
    if (options?.endDate) params.append('endDate', options.endDate.toISOString());
    if (options?.limit) params.append('limit', options.limit.toString());
    if (options?.offset) params.append('offset', options.offset.toString());
    if (options?.sortBy) params.append('sortBy', options.sortBy);
    if (options?.sortOrder) params.append('sortOrder', options.sortOrder);

    const response = await this.http.get<{
      transactions: TransactionRecord[];
      total: number;
      hasMore: boolean;
    }>(`/tx/history?${params.toString()}`);

    return response.data;
  }

  /**
   * Cancel a pending transaction
   */
  async cancel(id: string, reason?: string): Promise<TransactionRecord> {
    const response = await this.http.post<TransactionRecord>(`/tx/${id}/cancel`, { reason });
    
    this.events.emit('transaction:cancelled', { id, reason });

    return response.data;
  }

  /**
   * Retry a failed transaction
   */
  async retry(id: string): Promise<TransactionRecord> {
    const response = await this.http.post<TransactionRecord>(`/tx/${id}/retry`);
    
    this.events.emit('transaction:retried', { id });

    return response.data;
  }

  /**
   * Get transaction status updates
   */
  async getStatusUpdates(id: string): Promise<{
    updates: {
      status: TransactionStatus;
      timestamp: string;
      message?: string;
    }[];
  }> {
    const response = await this.http.get(`/tx/${id}/status-updates`);
    return response.data;
  }

  /**
   * Request expedited processing
   */
  async expedite(id: string): Promise<{
    success: boolean;
    estimatedCompletionTime?: string;
    additionalFee?: string;
  }> {
    const response = await this.http.post(`/tx/${id}/expedite`);
    return response.data;
  }

  /**
   * Get transaction receipt
   */
  async getReceipt(id: string): Promise<{
    transactionHash: string;
    blockNumber: number;
    blockHash: string;
    gasUsed: string;
    effectiveGasPrice: string;
    status: boolean;
    logs: any[];
  }> {
    const response = await this.http.get(`/tx/${id}/receipt`);
    return response.data;
  }

  /**
   * Estimate transaction cost
   */
  async estimateCost(request: TransactionRequest): Promise<{
    gasEstimate: string;
    gasPriceWei: string;
    totalCostWei: string;
    totalCostEth: string;
    totalCostUsd?: string;
  }> {
    const response = await this.http.post('/tx/estimate', request);
    return response.data;
  }

  /**
   * Get pending transactions for an address
   */
  async getPending(address: string): Promise<TransactionRecord[]> {
    const response = await this.http.get<{ transactions: TransactionRecord[] }>(
      `/tx/pending/${address}`
    );
    return response.data.transactions;
  }
}
